### PR TITLE
CP-502: Add confluent control center

### PR DESCRIPTION
**Pull request by kkonstantine**

### CODE REVIEW

**Comment by kkonstantine** [comment]
Requires `control-center-dev.properties` from https://github.com/confluentinc/blueway/pull/1611
to be under `./etc/confluent-control-center`. (might have to install manually before this PR is merged and packaged)

**Comment by ybyzek** [comment]
Hi Konstantine, I'm concerned that the PR as it is today just starts C3 service but doesn't make it usable with the rest of the services:

1. Can I get the code to test it out?
2. By bringing up C3 with confluent cli, will the kafka broker be automatically added to the brokers view?  I don't think so unless there's another PR that modifies server.properties to include the required Confluent Metrics Reporter configuration parameters. 
3. By bringing up C3 with confluent cli, will connect be automatically added to the connect management view?
4. Does it set replication to 1 for all four C3 topics?  (the default is replication=3)

**Comment by norwood** [comment]
@ybyzek 
you can just check this pr out if you want to test it.
yes, the pr that konstantine linked to changes the replication to 1
no, it doesnt do that. i dont know who owns that sort of thing, maybe @confluentinc/core ?

**Comment by ybyzek** [comment]
@norwood , thanks for pointing out the PR for control-center-dev.properties, I missed that link

IMO we also must have the change that enables Confluent Metrics Reporters in the broker -- {{server-dev.properties}} ?   We need it for two reasons:

1. Show value of C3.  We need the local kafka server there at minimum
2. Avoid confusion.  Users will say "I started C3 but I don't see my broker..."

If we don't have that PR that enables Confluent Metrics Reporters in the broker, then I almost would say let's NOT merge this in...just to avoid confusion.

**Comment by kkonstantine** [comment]
@ybyzek added method to enable metrics reporter and interceptors in the appropriate services when in enterprise mode.

File: src/oss/confluent.sh [hunk 1]
```diff
@@ -686,6 +755,43 @@ config_service() {
...
+    is_enterprise
+    status=$?
+    if [[ ${status} -ne 0 ]]; then
+        return 1
+    fi
+
+    local service="${1}"
+
+    local service_dir="${confluent_current}/${service}"
+    echo "metric.reporters=io.confluent.metrics.reporter.ConfluentMetricsReporter" \
```
Comment (ybyzek) [inline src/oss/confluent.sh]: I think C3 has two properties file, one of which is `-dev`.  Is there value in having the same thing for `server.properties`?  ie, `server-dev.properties`?

File: src/oss/confluent.sh [hunk 2]
```diff
@@ -615,15 +638,61 @@ wait_ksql() {
     wait_process_up "${pid}" 2000 || echo "ksql failed to start"
 }
 
+start_control-center() {
+    local service="control-center"
+    is_running "connect" "false" \
+        || die "Cannot start Control-Center, Kafka Connect is not running. Check your deployment"
```
Comment (ybyzek) [inline src/oss/confluent.sh]: Is this what we want?  ie, if Connect hasn't started, does that mean C3 should fail?  I thought C3 could still come up...but not sure.  cc: @norwood

File: src/oss/confluent.sh [hunk 3]
```diff
@@ -686,6 +755,43 @@ config_service() {
     fi
 }
 
+enable_metrics_reporter() {
```
Comment (ybyzek) [inline src/oss/confluent.sh]: Will `enable_metrics_reporter` be run every single time confluent cli starts services?  If yes, then would the echo statements below be redundant if it's run more than once?  ie, the properties file will have each line written to it each time?

File: src/oss/confluent.sh [hunk 4]
```diff
@@ -686,6 +755,43 @@ config_service() {
...
+    echo "metric.reporters=io.confluent.metrics.reporter.ConfluentMetricsReporter" \
+        >> "${service_dir}/${service}.properties"
+    echo "confluent.metrics.reporter.bootstrap.servers=localhost:${kafka_port}" \
+        >> "${service_dir}/${service}.properties"
+    echo "confluent.metrics.reporter.topic.replicas=1" >> "${service_dir}/${service}.properties"
+
+    return 0
+}
+
+enable_monitoring_interceptors() {
```
Comment (ybyzek) [inline src/oss/confluent.sh]: Same as above, Will `enable_monitoring_interceptors` be run every single time confluent cli starts services? If yes, then would the echo statements below be redundant if it's run more than once? ie, the properties file will have each line written to it each time?

File: src/oss/confluent.sh (hunk 2 above)
Comment (xvrl) [reply src/oss/confluent.sh]: connect is optional for c3

Comment (kkonstantine) [reply src/oss/confluent.sh]: Currently, given that the startup sequence is only one and linear this doesn't really matter. `control-center` comes after `connect` anyways, and allowing to run with connect down is not very useful. All that this means is that something was wrong with connect but C3 doesn't care. 

I'll change the dependency to be `kafka`. But again, with `confluent start` all the services should be up (and currently `confluent start control-center` is synonymous to `confluent start` with control-center being the last service in the startup order.

File: src/oss/confluent.sh (hunk 1 above)
Comment (kkonstantine) [reply src/oss/confluent.sh]: I believe such generalization in packaging gets out of the scope of this PR. The other services work sufficiently with their current configs. My suggestion is that we should reconsider such a change in a later release.

File: src/oss/confluent.sh (hunk 3 above)
Comment (kkonstantine) [reply src/oss/confluent.sh]: The properties under `etc` are the source of truth and are not overwritten. 

The properties under `$CONFLUENT_CURRENT` are the ones that are used to start the services with minor changes/additions to the equivalent file in `etc` (such as the one introduced here) and are overwritten completely with every run. 

So there's no problem with re-runs. A new file is created every time.

File: src/oss/confluent.sh (hunk 4 above)
Comment (kkonstantine) [reply src/oss/confluent.sh]: Same as with metrics reporter.

File: src/oss/confluent.sh (hunk 3 above)
Comment (ybyzek) [inline src/oss/confluent.sh]: Also, is this done for COS?  From what I can tell, it is.  But I think we may want it only for CE.

Comment (kkonstantine) [reply src/oss/confluent.sh]: It is not. In both functions added, there's a call to `is_enterprise` at the beginning to enable this appropriately.

Comment (ybyzek) [reply src/oss/confluent.sh]: I see that now, I'm sorry I missed it.  It is very obvious now ;)

**Comment by kkonstantine** [comment]
Thanks for the reviews everyone! We seem good to go

**Pull request merged**